modeshape-jcr/src/main/java/org/modeshape/jcr/spi/index/provider/IndexProvider.java

```diff
@@ -630,11 +739,82 @@ private boolean matchesType( IndexDefinition defn,
+        }
+        
+        // Does this index apply to any of the ANDed constraints?
+        int costEstimate = getCostEstimate();
+        for (Constraint constraint : calculator.andedConstraints()) {
+            if (planner.indexAppliesTo(constraint)) {
+                logger().trace("Index '{0}' in '{1}' provider applies to query in workspace '{2}' with constraint: {3}",
+                               defn.getName(), getName(), workspaceName, constraint);
+                // The index does apply to this constraint ...
+                long cardinality = index.estimateCardinality(constraint, context.getVariables());
+                long total = index.estimateTotalCount();
+                Float selectivity = null;
+                if (total > 0L) {
+                    double ratio = (double)cardinality / (double)total;
+                    selectivity = cardinality <= total ? new Float(ratio) : IndexCostCalculator.MAX_SELECTIVITY;
+                }
+                calculator.addIndex(defn.getName(), workspaceName, getName(), Collections.singleton(constraint),
+                                    costEstimate,
+                                    cardinality, selectivity);
+            }
+        }
+
+        // Does this index apply to any of the join conditions ...
+        for (JoinCondition joinCondition : calculator.joinConditions()) {
+            if (planner.indexAppliesTo(joinCondition)) {
+                logger().trace("Index '{0}' in '{1}' provider applies to query in workspace '{2}' with constraint: {3}",
+                               defn.getName(), getName(), workspaceName, joinCondition);
+                // The index does apply to this constraint, but the number of values corresponds to the total number of values
+                // in the index (this is a JOIN CONDITON for which there is no literal values) ...
+                long total = index.estimateTotalCount();
+                calculator.addIndex(defn.getName(), workspaceName, getName(), Collections.singleton(joinCondition),
+                                    costEstimate,
+                                    total);
+            }
+
+        }
+    }
+
+    /**
+     * Returns an object which is used during the planning phase to evaluate if a certain index should be used or not.
+     *
+     * @param context the context of the original query; never null
+     * @param calculator the calculator that should be used to record plan information for the index; never null
+     * @param defn the definition for the index; never null
+     * @return an {@link IndexUsage} instance, never null.
+     * 
+     * @see #planUseOfIndex(QueryContext, IndexCostCalculator, String, ManagedIndex, IndexDefinition) 
+     */
+    protected IndexUsage evaluateUsage( QueryContext context,
+                                        IndexCostCalculator calculator,
+                                        IndexDefinition defn ) {
+        return null;
```

Comment:
Line 795 says this never returns null. Should this be made abstract rather than return an invalid value by default?